src/lib/optimizer/strategy/chunk_pruning_rule.cpp

```diff
@@ -21,150 +22,156 @@
 
 namespace opossum {
 
-void ChunkPruningRule::_apply_to_plan_without_subqueries(const std::shared_ptr<AbstractLQPNode>& lqp_root) const {
-  // we only want to follow chains of predicates
-  if (lqp_root->type != LQPNodeType::Predicate) {
-    _apply_to_plan_inputs_without_subqueries(lqp_root);
-    return;
-  }
-
-  DebugAssert(lqp_root->input_count() == 1, "Predicate nodes should only have 1 input");
-  // try to find a chain of predicate nodes that ends in a leaf
-  std::vector<std::shared_ptr<PredicateNode>> predicate_nodes;
-
-  // Gather PredicateNodes on top of a StoredTableNode. Ignore non-filtering and ValidateNodes.
-  auto current_node = lqp_root;
-  while (current_node->type == LQPNodeType::Predicate || current_node->type == LQPNodeType::Validate ||
-         _is_non_filtering_node(*current_node)) {
-    // Once a node has multiple outputs, we cannot use the predicate nodes above any more. Otherwise, we might prune
-    // based on the conditions found only in a single branch.
-    if (current_node->output_count() > 1) {
-      predicate_nodes.clear();
-    }
+using PredicateChain = std::vector<std::shared_ptr<PredicateNode>>;
 
-    if (current_node->type == LQPNodeType::Predicate) {
-      predicate_nodes.emplace_back(std::static_pointer_cast<PredicateNode>(current_node));
-    }
+void ChunkPruningRule::_apply_to_plan_without_subqueries(const std::shared_ptr<AbstractLQPNode>& lqp_root) const {
+  std::unordered_map<std::shared_ptr<StoredTableNode>, std::vector<PredicateChain>>
+      predicate_chains_by_stored_table_node;
 
-    current_node = current_node->left_input();
+  // (1) Collect StoredTableNodes
+  const auto stored_table_nodes = lqp_find_leafs<StoredTableNode>(lqp_root);
+  for (const auto& stored_table_node : stored_table_nodes) {
+    predicate_chains_by_stored_table_node.emplace(stored_table_node, std::vector<PredicateChain>{});
   }
 
-  if (current_node->type != LQPNodeType::StoredTable) {
-    _apply_to_plan_inputs_without_subqueries(lqp_root);
-    return;
-  }
-  const auto stored_table = std::static_pointer_cast<StoredTableNode>(current_node);
-  DebugAssert(stored_table->input_count() == 0, "Stored table nodes should not have inputs.");
-
-  /**
-   * A chain of predicates followed by a stored table node was found.
-   */
-  auto table = Hyrise::get().storage_manager.get_table(stored_table->table_name);
-
-  std::set<ChunkID> pruned_chunk_ids;
-  for (auto& predicate : predicate_nodes) {
-    auto new_exclusions = _compute_exclude_list(*table, *predicate->predicate(), stored_table);
-    pruned_chunk_ids.insert(new_exclusions.begin(), new_exclusions.end());
+  // (2) Collect chains of PredicateNodes on top of each StoredTableNode
+  for (auto& [stored_table_node, predicate_chains] : predicate_chains_by_stored_table_node) {
+    predicate_chains = _find_predicate_chains_recursively(stored_table_node, stored_table_node);
```

Comment:
I mentioned the interface of this method before. This line is somewhat hard to understand if you do not know the implementation of the method.